DESIGN_GUIDELINES.md

```diff
@@ -1165,6 +1226,87 @@ use the SecretValue type [_awslint:secret-token_].
+kinds of resources. For example, the
+`aws-ecs-patterns.ApplicationLoadBalancedFargateService` construct represents an
+architecture that includes an AWS Fargate container cluster employing an
+Application Load Balancer (ALB). These examples are rare, as these patterns
+constructs are, by design, heavily opinionated and less customizable; this means
+while the patterns are useful for many customers as-is, many others will find
+the patterns not _quite_ right and inflexible to their needs.
+
+#### What Belongs in the CDK?
+
+What divides the L2s from the higher-level constructs, and where do we draw the
+line on what belongs as part of the CDK itself?
+
+Generally speaking, an L2 directly models a single CloudFormation resource. The
+L2 typically exposes all of the underlying surface area of the underlying L1, as
+well as provides convenience methods for interactions with other services, or
+creating related or dependent resources, permissions, and metrics.
+
+Examples of behaviors that an L2 commonly include:
+
+- Modeling of the underlying L1 properties
+- Methods for adding related resources (e.g., adding an event notification to
+  an S3 bucket).
+- Modeling of permissions and resource policies
+- Modeling of metrics
+
+In addition to the above, some higher-order L2s may introduce more complex and
+helpful functionality, either part of the original L2 itself, or as part of a
+separate construct. The most common form of these L2s are integration constructs
+that model interactions between different services (e.g., SNS publishing to SQS,
+CodePipeline actions that trigger Lambda functions).
+
+A higher-level L2 may also be included in the CDK if another L2 requires or
+would greatly benefit from a dependency on it.
+
+As general guidelines, the following types of constructs likely don't belong in
+the CDK, and should be published as separate artifacts:
+
+- Constructs that create complete application infrastructure spanning multiple
+  resources (e.g., networking + compute). These constructs are often very
+  helpful to a specific subset of customers, but require high levels of
+  customization to work for all customers.
+- Constructs that extend the base L2s to provide functionality for a specific
+  language (e.g., `PythonFunction`), operating system (e.g., `WindowsInstance`),
+  or framework (e.g., `FlaskRestApi`). Rationale: subclassing the L2s in the CDK
+  leads to awkward inheritence chains, especially for customers who themselves
+  subclass the constructs to enforce organizational standards (e.g., `Bucket` ->
+  `SecureBucket`). Prefer to integrate this support directly into the
+  base L2 -- either directly or via composition -- where possible.
+- Constructs which assume a particular organizational, team, or architectural
+  structure. An example here might be a construct for validating
+  CertificateManager certificates which presumes a specific account hierarchy
```

Comment:
I am not sure that's the best example. If there is a very common need, it may make sense to include something like that in the library.

Reply:
Fair. The key words here are "very common"; if a structure is particular to one (or even a few) organizations, that likely doesn't meet the bar for "very common" usage. That's what I was trying to capture here. Suggestions for a better example?

Reply:
I feel like we should focus on defining what goes in and not spend too much verbiage for what stays out. The sky is the limit when it comes to abstractions on top of the CDK so I am not sure it adds much value to provide a few semi-random examples. 

If we have good tenets (or at least a decision-making model) for determining if a certain api should be included in the CDK itself, all the rest stays out by definition.

Reply:
Dropped "what doesn't fit" section.